ipa-core/src/net/test.rs

```diff
@@ -140,31 +251,75 @@ impl TestConfigBuilder {
         self
     }
 
-    #[must_use]
-    pub fn build(self) -> TestConfig {
-        let mut sockets = None;
-        let ports = self.ports.unwrap_or_else(|| {
-            let socks = array::from_fn(|_| TcpListener::bind("localhost:0").unwrap());
-            let ports = socks
-                .each_ref()
-                .map(|sock| sock.local_addr().unwrap().port());
-            sockets = Some(socks);
-            ports
+    /// Creates 3 MPC test servers, assigning ports from OS if necessary.
+    fn build_mpc_servers(
+        &self,
+        optional_ports: Option<Vec<u16>>,
+        ring_index: usize,
+    ) -> TestServers {
+        let mut sockets = vec![None, None, None];
+        let ports = optional_ports.unwrap_or_else(|| {
+            sockets = (0..3)
+                .map(|_| Some(TcpListener::bind("localhost:0").unwrap()))
+                .collect();
+            sockets
+                .iter()
+                .map(|sock| sock.as_ref().unwrap().local_addr().unwrap().port())
+                .collect()
         });
-        let (scheme, certs) = if self.disable_https {
-            ("http", [None, None, None])
+        // Ports will always be defined after this. Socks only if there were no ports set.
+        let configs = if self.disable_https {
+            ports
+                .into_iter()
+                .map(|ports| server_config_insecure_http(ports, !self.disable_matchkey_encryption))
+                .collect()
         } else {
-            ("https", TEST_CERTS_DER.clone().map(Some))
+            let start_idx = ring_index * 3;
+            (start_idx..start_idx + 3)
+                .map(|id| server_config_https(id, ports[id], !self.disable_matchkey_encryption))
+                .collect()
         };
-        let peers = certs
-            .into_iter()
-            .enumerate()
-            .map(|(i, cert)| PeerConfig {
-                url: format!("{scheme}://localhost:{}", ports[i])
-                    .parse()
-                    .unwrap(),
-                certificate: cert,
-                hpke_config: if self.disable_matchkey_encryption {
+        TestServers::new(configs, sockets)
+    }
+
+    /// Defines a test network with a single MPC ring (3 helpers).
+    fn build_ring_network(
```

Comment:
After this change `TestConfig` looks a lot more complicated that it was before. I think it will benefit from some refactoring to split MPC and shard network initializations

Reply:
Yeah, Test config holds now the entire sharded network. But on the other hand, having everything described on this single object makes it easy to operate on the network as a whole which is the use case (check the transport.rs test). Separating this into 2 seems like it will only add complexity on the caller side.

Can you be more explicit on how you want to split the network init?

Reply:
you don't have to make callers suffer. Providing a single entry point for them to construct the config is reasonable. Doing everything inside one struct after that is not necessary - power through composition